src/v/cloud_storage/remote_partition.cc

```diff
@@ -174,7 +174,11 @@ remote_partition::borrow_result_t remote_partition::borrow_next_segment_reader(
     auto iter = _segments.find(mit->base_offset);
     if (iter != _segments.end()) {
         if (
-          iter->second->segment->get_max_rp_offset() != mit->committed_offset) {
+          iter->second->segment->get_max_rp_offset() != mit->committed_offset
+          || iter->second->segment->get_segment_size() != mit->size_bytes) {
+            // The segment was replaced and doesn't match metadata anymore. We
+            // want to avoid picking it up because otherwise we won't be able to
+            // make any progress.
```

Comment:
maybe to future proof this: 

```suggestion
if(iter->second->segment->get_segment_path()!=manifest.generate_segment_path(*mit)) {
   offload_segment(iter->first);
   iter = _segments.end();
}
```

or since this is relatively expensive operation, we could extract into a class the fields that make up the remote identity of the segment (ntp, rev, base, committed, size,seg term, name_version,  archiver_term) and compare this

Reply:
done